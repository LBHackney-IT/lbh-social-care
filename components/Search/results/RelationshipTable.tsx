import Link from 'next/link';
import cx from 'classnames';
import { LegacyResident } from 'types';

const ResultEntry = ({
  person,
  newTab,
  getValueCallback,
}: {
  person: LegacyResident;
  newTab?: boolean;
}): React.ReactElement => {
  const { mosaicId, firstName, lastName, dateOfBirth, address } = person;

  return (
    <>
      <tr className={cx('govuk-table__row')}>
        <td className="govuk-table__cell">
          <input
            aria-labelledby={`worker_${mosaicId}`}
            name="workerId"
            type="radio"
            value={mosaicId}
            onChange={(e) => {
              getValueCallback(e.target.value);
            }}
          />
        </td>
        <td className="govuk-table__cell">{mosaicId}</td>
        <td className="govuk-table__cell">
          {newTab ? (
            <a
              href={`/people/${mosaicId}`}
              target="_blank"
              rel="noreferrer"
              className="govuk-link govuk-custom-text-color"
            >
              {firstName} {lastName}
            </a>
          ) : (
            <Link href={`/people/${mosaicId}`}>
              <a className="govuk-link govuk-custom-text-color">
                {firstName} {lastName}
              </a>
            </Link>
          )}
        </td>
        <td className="govuk-table__cell">
          <span>{(address && address.postcode) || ''}</span>
        </td>
        <td className="govuk-table__cell">
          {dateOfBirth && new Date(dateOfBirth).toLocaleDateString('en-GB')}
        </td>
      </tr>
    </>
  );
};

const RelationshipSearchTable = ({
  records,
  getValueCallback,
}: {
  records: LegacyResident[];
}): React.ReactElement => {
  return (
    <table className="govuk-table lbh-table" data-testid="residents-table">
      <thead className="govuk-table__head">
        <tr className="govuk-table__row">
          <th scope="col" className="govuk-table__header">
            Select
          </th>
          <th scope="col" className="govuk-table__header">
            ID
          </th>
          <th scope="col" className="govuk-table__header">
            Client Name
          </th>
          <th scope="col" className="govuk-table__header">
            Address
          </th>
          <th scope="col" className="govuk-table__header">
            Date of birth
          </th>
          <th scope="col" className="govuk-table__header" />
        </tr>
      </thead>

      <tbody className="govuk-table__body">
        {records.map((result) => (
          <ResultEntry
            key={result.mosaicId}
            person={result}
            getValueCallback={getValueCallback}
          />
        ))}
      </tbody>
    </table>
  );
};

export default RelationshipSearchTable;
