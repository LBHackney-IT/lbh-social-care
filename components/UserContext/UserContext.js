import { createContext, useContext, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { AUTH_WHITELIST } from 'utils/auth';
import { getUser } from 'utils/api/checkAuth';
import Layout from 'components/Layout';
import Spinner from 'components/Spinner/Spinner';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState();
  const [redirect, setRedirect] = useState();
  const [isLoading, setLoading] = useState(true);
  const router = useRouter();
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        const data = getUser();
        setUser(data);
      } catch (e) {
        setUser();
        setRedirect(e.response.status === 403 ? '/access-denied' : '/login');
      }
      setLoading(false);
    };
    initializeAuth();
  }, []);
  if (isLoading || !router.isReady) {
    return (
      <Layout>
        <Spinner />
      </Layout>
    );
  }
  if (
    typeof window !== 'undefined' &&
    AUTH_WHITELIST.includes(router.pathname) === Boolean(user)
  ) {
    router.push(redirect || location);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
        isLoading,
        setUser,
        redirect,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
