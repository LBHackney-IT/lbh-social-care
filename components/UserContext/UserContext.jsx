import { createContext, useContext } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children, user }) => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
