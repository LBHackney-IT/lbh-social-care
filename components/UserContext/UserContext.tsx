import { createContext, useContext, ReactNode, ReactElement } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

import type { User } from 'types';

interface Props {
  children: ReactNode;
  user?: User;
}

export const UserContext = createContext<{ user?: User }>({
  user: undefined,
});

export const AuthProvider = ({ children, user }: Props): ReactElement => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export const useAuth = (): { user?: User } => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
