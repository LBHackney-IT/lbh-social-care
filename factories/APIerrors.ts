import { Factory } from 'fishery';

import { ErrorAPI } from 'types';

export const errorFactory = Factory.define<ErrorAPI>(() => ({
  response: {
    data: {},
    status: 404,
    statusText: 'error',
    headers: {},
    config: {},
  },
  name: 'error',
  message: 'error',
  config: {},
  isAxiosError: false,
  toJSON: () => ({}),
}));

export const mockedAPIerror = errorFactory.build();
export const mockedAPIservererror = errorFactory.build({
  response: {
    status: 500,
  },
});
