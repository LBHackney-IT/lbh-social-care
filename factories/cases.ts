import { Factory } from 'fishery';

import { Case } from 'types';

const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'ASC_case_note',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
}));

const allocationCaseFactory = caseFactory.params({
  formName: 'Worker allocated',
  caseFormUrl: undefined,
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});

const deallocationCaseFactory = caseFactory.params({
  formName: 'Worker allocated',
  caseFormUrl: undefined,
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = allocationCaseFactory.build();
export const mockedDeallocationNote = deallocationCaseFactory.build();
