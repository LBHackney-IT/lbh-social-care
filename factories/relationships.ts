import { Factory } from 'fishery';
import { RelationshipData, RelationshipPerson } from 'types';

const mockedRelationshipFactory = Factory.define<RelationshipData>(
  ({ sequence }) => ({
    personId: sequence,
    personalRelationships: {
      parents: [],
      siblings: [],
      children: [],
      other: [],
    },
  })
);

const mockedRelationPerson = Factory.define<RelationshipPerson>(
  ({ sequence }) => ({
    id: sequence,
    firstName: 'mock_me',
    lastName: 'mock_me',
  })
);

export const mockedRelationship = mockedRelationshipFactory.build({
  personId: 123,
  personalRelationships: {
    parents: [
      mockedRelationPerson.build({
        id: 111,
        firstName: 'Giovanni',
        lastName: 'Muciaccia',
      }),
    ],

    siblings: [
      mockedRelationPerson.build({
        id: 222,
        firstName: 'Andrea',
        lastName: 'Loprinzo',
      }),
    ],

    children: [
      mockedRelationPerson.build({
        id: 333,
        firstName: 'Giuseppe',
        lastName: 'Geppetto',
      }),
    ],

    other: [
      mockedRelationPerson.build({
        id: 444,
        firstName: 'Pinocchio',
        lastName: 'Geppetto',
      }),
    ],
  },
});

export const mockedRelationshipNoData = mockedRelationshipFactory.build({
  personId: 123,
  personalRelationships: {
    parents: [],
    siblings: [],
    children: [],
    other: [],
  },
});

export const mockRelationshipEmptyData = mockedRelationshipFactory.build({
  personId: 123,
  personalRelationships: {},
});

export const mockedRelationshipPartialData = mockedRelationshipFactory.build({
  personId: 123,
  personalRelationships: {
    parents: [
      mockedRelationPerson.build({
        id: 111,
        firstName: 'Mastro',
        lastName: 'Geppetto',
      }),
    ],
    siblings: [],
    children: [
      mockedRelationPerson.build({
        id: 123,
        firstName: 'Pinocchio',
        lastName: 'Geppetto',
      }),
    ],
    other: [],
  },
});
