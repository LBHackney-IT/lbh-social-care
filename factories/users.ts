import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
}));

export const mockedUser = userFactory.build();
