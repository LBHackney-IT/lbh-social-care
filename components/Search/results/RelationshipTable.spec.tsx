import { render } from '@testing-library/react';

import RelationshipTable from './RelationshipTable';
import {
  legacyResidentFactory,
  mockedLegacyResident,
} from 'factories/residents';
import { UserContext } from '../../UserContext/UserContext';
import { mockedUser, userFactory } from '../../../factories/users';

describe('Relationshiptable component', () => {
  it('should render a list of residents', () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ user: mockedUser }}>
        <RelationshipTable records={[mockedLegacyResident]} callback={null} />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a resident without a postcode', () => {
    const { asFragment } = render(
      <UserContext.Provider
        value={{
          user: userFactory.build({
            hasAdminPermissions: false,
            hasDevPermissions: false,
            hasChildrenPermissions: false,
            hasUnrestrictedPermissions: false,
            hasAllocationsPermissions: false,
            hasAdultPermissions: true,
          }),
        }}
      >
        <RelationshipTable
          records={[
            legacyResidentFactory.build({
              ageContext: 'C',
              address: {
                address: 'some address',
                postcode: null,
              },
            }),
          ]}
          callback={null}
        />
      </UserContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
