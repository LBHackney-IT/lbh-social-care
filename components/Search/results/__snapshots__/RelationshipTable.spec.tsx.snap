// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Relationshiptable component should render a list of residents 1`] = `
<DocumentFragment>
  <table
    class="govuk-table lbh-table"
    data-testid="residents-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Select
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          ID
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Client Name
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Address
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          <input
            aria-labelledby="person_1"
            name="personId"
            type="radio"
            value="1"
          />
        </td>
        <td
          class="govuk-table__cell"
        >
          1
        </td>
        <td
          class="govuk-table__cell"
        >
          Foo Bar
        </td>
        <td
          class="govuk-table__cell"
        >
          <span />
        </td>
        <td
          class="govuk-table__cell"
        >
          13/11/2020
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Relationshiptable component should render a resident without a postcode 1`] = `
<DocumentFragment>
  <table
    class="govuk-table lbh-table"
    data-testid="residents-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Select
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          ID
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Client Name
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Address
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          <input
            aria-labelledby="person_2"
            name="personId"
            type="radio"
            value="2"
          />
        </td>
        <td
          class="govuk-table__cell"
        >
          2
        </td>
        <td
          class="govuk-table__cell"
        >
          Foo Bar
        </td>
        <td
          class="govuk-table__cell"
        >
          <span />
        </td>
        <td
          class="govuk-table__cell"
        >
          13/11/2020
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
